{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:16:54.022803Z",
     "iopub.status.busy": "2024-12-04T09:16:54.022407Z",
     "iopub.status.idle": "2024-12-04T09:16:55.633893Z",
     "shell.execute_reply": "2024-12-04T09:16:55.632413Z"
    }
   },
   "outputs": [],
   "source": [
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:16:55.638238Z",
     "iopub.status.busy": "2024-12-04T09:16:55.637763Z",
     "iopub.status.idle": "2024-12-04T09:16:55.645539Z",
     "shell.execute_reply": "2024-12-04T09:16:55.644524Z"
    }
   },
   "outputs": [],
   "source": [
    "# Ensure cwd is project root and that /code is in path\n",
    "\n",
    "import os\n",
    "import sys\n",
    "\n",
    "cwd = os.path.basename(os.getcwd())\n",
    "\n",
    "if cwd == 'footprintQTL':\n",
    "    \n",
    "    PROJECT_DIR = '.'\n",
    "\n",
    "\n",
    "elif cwd == 'code':\n",
    "    \n",
    "    PROJECT_DIR = '..'\n",
    "\n",
    "\n",
    "elif cwd == 'fichtner':\n",
    "\n",
    "    PROJECT_DIR = 'projects/footprintQTL'\n",
    "\n",
    "else:\n",
    "\n",
    "    PROJECT_DIR = 'manual'\n",
    "\n",
    "\n",
    "os.chdir(PROJECT_DIR)\n",
    "sys.path.append(os.getcwd() + '/code')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:16:55.648845Z",
     "iopub.status.busy": "2024-12-04T09:16:55.648544Z",
     "iopub.status.idle": "2024-12-04T09:17:01.674224Z",
     "shell.execute_reply": "2024-12-04T09:17:01.672479Z"
    }
   },
   "outputs": [],
   "source": [
    "from helpers.python.utils import ct_format, ct_format_alt, create_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:17:01.681147Z",
     "iopub.status.busy": "2024-12-04T09:17:01.680538Z",
     "iopub.status.idle": "2024-12-04T09:17:01.702250Z",
     "shell.execute_reply": "2024-12-04T09:17:01.701040Z"
    }
   },
   "outputs": [],
   "source": [
    "#User\n",
    "\n",
    "from glob_vars import GENOTYPES_PROCESSED_TSV, SNP_LOCS_BED, PRECOMPUTED_EQTLS_TSV, CT_MAP_JSON, CT_MAP_ID, EQTLS_DIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:17:01.711179Z",
     "iopub.status.busy": "2024-12-04T09:17:01.710700Z",
     "iopub.status.idle": "2024-12-04T09:17:01.719829Z",
     "shell.execute_reply": "2024-12-04T09:17:01.718706Z"
    }
   },
   "outputs": [],
   "source": [
    "# Get ct-ann --> grouped-ct mappings\n",
    "with open(CT_MAP_JSON, 'r') as f:\n",
    "    ct_map = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:17:01.723805Z",
     "iopub.status.busy": "2024-12-04T09:17:01.723425Z",
     "iopub.status.idle": "2024-12-04T09:17:01.735459Z",
     "shell.execute_reply": "2024-12-04T09:17:01.734165Z"
    }
   },
   "outputs": [],
   "source": [
    "# Format\n",
    "\n",
    "# ct_map = {ct_format(new_ct): [ct_format(old_ct) for old_ct in old_cts] for new_ct, old_cts in ct_map.items()}\n",
    "# ct_map_i = {old_ct: new_ct for new_ct, old_cts in ct_map.items() for old_ct in old_cts}\n",
    "\n",
    "ct_map_alt = {ct_format(new_ct): [ct_format_alt(old_ct) for old_ct in old_cts] for new_ct, old_cts in ct_map.items()}\n",
    "ct_map_i_alt = {old_ct: new_ct for new_ct, old_cts in ct_map_alt.items() for old_ct in old_cts}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotype tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:17:01.739652Z",
     "iopub.status.busy": "2024-12-04T09:17:01.739317Z",
     "iopub.status.idle": "2024-12-04T09:17:46.103261Z",
     "shell.execute_reply": "2024-12-04T09:17:46.101864Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMEA2536416</th>\n",
       "      <th>SAMEA3854313</th>\n",
       "      <th>SAMEA3853161</th>\n",
       "      <th>SAMEA3966392</th>\n",
       "      <th>SAMEA2625592</th>\n",
       "      <th>SAMEA2698309</th>\n",
       "      <th>SAMEA3962624</th>\n",
       "      <th>SAMEA3965165</th>\n",
       "      <th>SAMEA2595426</th>\n",
       "      <th>SAMEA3962389</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMEA2555017</th>\n",
       "      <th>SAMEA2420640</th>\n",
       "      <th>SAMEA4451109</th>\n",
       "      <th>SAMEA3448738</th>\n",
       "      <th>SAMEA2627141</th>\n",
       "      <th>SAMEA3485958</th>\n",
       "      <th>SAMEA2536413</th>\n",
       "      <th>SAMEA2464819</th>\n",
       "      <th>SAMEA2627567</th>\n",
       "      <th>SAMEA2645814</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10_100000235_C_T</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100000943_G_A</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100000979_T_C</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100002628_A_C</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100002875_A_G</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9_9999539_A_G</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9_99997250_C_T</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9_99998141_T_C</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9_99998283_G_C</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9_99999321_G_C</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3263990 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     SAMEA2536416  SAMEA3854313  SAMEA3853161  SAMEA3966392  \\\n",
       "snp                                                                           \n",
       "chr10_100000235_C_T           2.0           0.0           1.0           1.0   \n",
       "chr10_100000943_G_A           1.0           2.0           1.0           2.0   \n",
       "chr10_100000979_T_C           2.0           2.0           2.0           2.0   \n",
       "chr10_100002628_A_C           1.0           0.0           0.0           1.0   \n",
       "chr10_100002875_A_G           2.0           2.0           2.0           2.0   \n",
       "...                           ...           ...           ...           ...   \n",
       "chr9_9999539_A_G              2.0           2.0           2.0           2.0   \n",
       "chr9_99997250_C_T             2.0           1.0           1.0           2.0   \n",
       "chr9_99998141_T_C             2.0           2.0           2.0           2.0   \n",
       "chr9_99998283_G_C             2.0           1.0           1.0           2.0   \n",
       "chr9_99999321_G_C             0.0           1.0           1.0           1.0   \n",
       "\n",
       "                     SAMEA2625592  SAMEA2698309  SAMEA3962624  SAMEA3965165  \\\n",
       "snp                                                                           \n",
       "chr10_100000235_C_T           1.0           1.0           2.0           2.0   \n",
       "chr10_100000943_G_A           2.0           2.0           2.0           2.0   \n",
       "chr10_100000979_T_C           2.0           2.0           2.0           2.0   \n",
       "chr10_100002628_A_C           1.0           1.0           2.0           2.0   \n",
       "chr10_100002875_A_G           2.0           2.0           2.0           2.0   \n",
       "...                           ...           ...           ...           ...   \n",
       "chr9_9999539_A_G              2.0           2.0           2.0           2.0   \n",
       "chr9_99997250_C_T             2.0           2.0           2.0           0.0   \n",
       "chr9_99998141_T_C             2.0           2.0           2.0           1.0   \n",
       "chr9_99998283_G_C             2.0           2.0           2.0           0.0   \n",
       "chr9_99999321_G_C             2.0           1.0           2.0           2.0   \n",
       "\n",
       "                     SAMEA2595426  SAMEA3962389  ...  SAMEA2555017  \\\n",
       "snp                                              ...                 \n",
       "chr10_100000235_C_T           0.0           1.0  ...           2.0   \n",
       "chr10_100000943_G_A           2.0           2.0  ...           2.0   \n",
       "chr10_100000979_T_C           2.0           2.0  ...           2.0   \n",
       "chr10_100002628_A_C           0.0           1.0  ...           2.0   \n",
       "chr10_100002875_A_G           2.0           2.0  ...           2.0   \n",
       "...                           ...           ...  ...           ...   \n",
       "chr9_9999539_A_G              2.0           2.0  ...           1.0   \n",
       "chr9_99997250_C_T             2.0           1.0  ...           1.0   \n",
       "chr9_99998141_T_C             2.0           1.0  ...           2.0   \n",
       "chr9_99998283_G_C             2.0           1.0  ...           1.0   \n",
       "chr9_99999321_G_C             2.0           2.0  ...           1.0   \n",
       "\n",
       "                     SAMEA2420640  SAMEA4451109  SAMEA3448738  SAMEA2627141  \\\n",
       "snp                                                                           \n",
       "chr10_100000235_C_T           2.0           2.0           1.0           1.0   \n",
       "chr10_100000943_G_A           2.0           2.0           2.0           2.0   \n",
       "chr10_100000979_T_C           2.0           2.0           2.0           2.0   \n",
       "chr10_100002628_A_C           2.0           2.0           1.0           1.0   \n",
       "chr10_100002875_A_G           2.0           2.0           2.0           2.0   \n",
       "...                           ...           ...           ...           ...   \n",
       "chr9_9999539_A_G              2.0           2.0           1.0           2.0   \n",
       "chr9_99997250_C_T             1.0           1.0           1.0           2.0   \n",
       "chr9_99998141_T_C             2.0           2.0           2.0           2.0   \n",
       "chr9_99998283_G_C             1.0           1.0           1.0           2.0   \n",
       "chr9_99999321_G_C             1.0           2.0           2.0           0.0   \n",
       "\n",
       "                     SAMEA3485958  SAMEA2536413  SAMEA2464819  SAMEA2627567  \\\n",
       "snp                                                                           \n",
       "chr10_100000235_C_T           2.0           0.0           2.0           2.0   \n",
       "chr10_100000943_G_A           1.0           2.0           1.0           1.0   \n",
       "chr10_100000979_T_C           2.0           2.0           1.0           1.0   \n",
       "chr10_100002628_A_C           0.0           0.0           1.0           1.0   \n",
       "chr10_100002875_A_G           2.0           2.0           1.0           1.0   \n",
       "...                           ...           ...           ...           ...   \n",
       "chr9_9999539_A_G              2.0           2.0           2.0           2.0   \n",
       "chr9_99997250_C_T             1.0           1.0           0.0           0.0   \n",
       "chr9_99998141_T_C             2.0           1.0           1.0           2.0   \n",
       "chr9_99998283_G_C             1.0           1.0           0.0           0.0   \n",
       "chr9_99999321_G_C             2.0           1.0           2.0           2.0   \n",
       "\n",
       "                     SAMEA2645814  \n",
       "snp                                \n",
       "chr10_100000235_C_T           2.0  \n",
       "chr10_100000943_G_A           2.0  \n",
       "chr10_100000979_T_C           2.0  \n",
       "chr10_100002628_A_C           2.0  \n",
       "chr10_100002875_A_G           2.0  \n",
       "...                           ...  \n",
       "chr9_9999539_A_G              1.0  \n",
       "chr9_99997250_C_T             2.0  \n",
       "chr9_99998141_T_C             2.0  \n",
       "chr9_99998283_G_C             2.0  \n",
       "chr9_99999321_G_C             2.0  \n",
       "\n",
       "[3263990 rows x 85 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt = pd.read_csv(GENOTYPES_PROCESSED_TSV, sep='\\t', header=0, index_col=0)\n",
    "gt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### snp_locations.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:17:46.169885Z",
     "iopub.status.busy": "2024-12-04T09:17:46.169299Z",
     "iopub.status.idle": "2024-12-04T09:18:00.210975Z",
     "shell.execute_reply": "2024-12-04T09:18:00.209707Z"
    }
   },
   "outputs": [],
   "source": [
    "snps_bed = gt.index.to_frame().reset_index(drop=True)\n",
    "snps_bed['contig'] = snps_bed['snp'].str.split('_').str[0].astype(str)\n",
    "\n",
    "# Assumed only chroms in contigs\n",
    "chroms = list(['chr' + str(i) for i in range(1,24)] + ['X', 'Y'])\n",
    "assert np.isin(snps_bed['contig'].unique(), chroms).all()\n",
    "\n",
    "# Format bed\n",
    "snps_bed['chr'] = snps_bed['contig']\n",
    "snps_bed['start'] = snps_bed['snp'].str.split('_').str[1].astype(int) - 1 # 1-based fully closed (GT matrix) to 0-based half open (bed tools and matrix-eQTL)\n",
    "snps_bed['end'] = snps_bed['snp'].str.split('_').str[1]\n",
    "\n",
    "snps_bed = snps_bed[['chr', 'start', 'end', 'snp']].sort_values(['chr', 'start'], ascending=[True, True])\n",
    "\n",
    "# Save bed\n",
    "snps_bed.to_csv(SNP_LOCS_BED, sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pre-computed eQTLs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make eQTLs bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:18:00.218057Z",
     "iopub.status.busy": "2024-12-04T09:18:00.217447Z",
     "iopub.status.idle": "2024-12-04T09:18:00.298455Z",
     "shell.execute_reply": "2024-12-04T09:18:00.297466Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>p_value</th>\n",
       "      <th>beta</th>\n",
       "      <th>beta_se</th>\n",
       "      <th>empirical_feature_p_value</th>\n",
       "      <th>feature_chromosome</th>\n",
       "      <th>feature_start</th>\n",
       "      <th>feature_end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>...</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>assessed_allele</th>\n",
       "      <th>call_rate</th>\n",
       "      <th>maf</th>\n",
       "      <th>hwe_p</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>global_corrected_pValue</th>\n",
       "      <th>global_corrected_pValue_BH</th>\n",
       "      <th>global_corrected_pValue_BF</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>QTL</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000269376-chr13_112983450_T_C</th>\n",
       "      <td>chr13_112983450_T_C</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>-0.364427</td>\n",
       "      <td>0.092665</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13</td>\n",
       "      <td>113009671</td>\n",
       "      <td>113010319</td>\n",
       "      <td>NaN</td>\n",
       "      <td>277</td>\n",
       "      <td>...</td>\n",
       "      <td>112983450</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.044776</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ENSG00000269376</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>immature_EN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183463-chr13_27824764_C_T</th>\n",
       "      <td>chr13_27824764_C_T</td>\n",
       "      <td>0.047441</td>\n",
       "      <td>-0.089739</td>\n",
       "      <td>0.045430</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13</td>\n",
       "      <td>27977717</td>\n",
       "      <td>27988693</td>\n",
       "      <td>URAD</td>\n",
       "      <td>277</td>\n",
       "      <td>...</td>\n",
       "      <td>27824764</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.089552</td>\n",
       "      <td>0.073438</td>\n",
       "      <td>ENSG00000183463</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>immature_EN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000259182-chr15_101221494_A_C</th>\n",
       "      <td>chr15_101221494_A_C</td>\n",
       "      <td>0.202509</td>\n",
       "      <td>-0.088910</td>\n",
       "      <td>0.069866</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>15</td>\n",
       "      <td>101168530</td>\n",
       "      <td>101170821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>277</td>\n",
       "      <td>...</td>\n",
       "      <td>101221494</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.044776</td>\n",
       "      <td>0.110206</td>\n",
       "      <td>ENSG00000259182</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>immature_EN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286922-chr12_127567286_A_G</th>\n",
       "      <td>chr12_127567286_A_G</td>\n",
       "      <td>0.265404</td>\n",
       "      <td>-0.030918</td>\n",
       "      <td>0.027792</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>12</td>\n",
       "      <td>127486938</td>\n",
       "      <td>127533242</td>\n",
       "      <td>NaN</td>\n",
       "      <td>277</td>\n",
       "      <td>...</td>\n",
       "      <td>127567286</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ENSG00000286922</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>immature_EN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286246-chr12_129375924_C_G</th>\n",
       "      <td>chr12_129375924_C_G</td>\n",
       "      <td>0.949835</td>\n",
       "      <td>-0.000101</td>\n",
       "      <td>0.001608</td>\n",
       "      <td>4.101945e-163</td>\n",
       "      <td>12</td>\n",
       "      <td>129622929</td>\n",
       "      <td>129625366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>277</td>\n",
       "      <td>...</td>\n",
       "      <td>129375924</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.089552</td>\n",
       "      <td>0.073438</td>\n",
       "      <td>ENSG00000286246</td>\n",
       "      <td>1.930191e-161</td>\n",
       "      <td>2.223254e-161</td>\n",
       "      <td>1.111627e-160</td>\n",
       "      <td>immature_EN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000250842-chr5_145190494_A_T</th>\n",
       "      <td>chr5_145190494_A_T</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.535958</td>\n",
       "      <td>0.130797</td>\n",
       "      <td>1.828702e-02</td>\n",
       "      <td>5</td>\n",
       "      <td>145337932</td>\n",
       "      <td>145381670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>310</td>\n",
       "      <td>...</td>\n",
       "      <td>145190494</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.316176</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ENSG00000250842</td>\n",
       "      <td>4.610052e-02</td>\n",
       "      <td>4.610052e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000267537-chr19_28873752_G_A</th>\n",
       "      <td>chr19_28873752_G_A</td>\n",
       "      <td>0.000285</td>\n",
       "      <td>-0.620958</td>\n",
       "      <td>0.172965</td>\n",
       "      <td>1.854836e-02</td>\n",
       "      <td>19</td>\n",
       "      <td>28602379</td>\n",
       "      <td>28648303</td>\n",
       "      <td>NaN</td>\n",
       "      <td>310</td>\n",
       "      <td>...</td>\n",
       "      <td>28873752</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.066176</td>\n",
       "      <td>0.246147</td>\n",
       "      <td>ENSG00000267537</td>\n",
       "      <td>4.669823e-02</td>\n",
       "      <td>4.669823e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000287056-chr6_168100952_C_G</th>\n",
       "      <td>chr6_168100952_C_G</td>\n",
       "      <td>0.001926</td>\n",
       "      <td>-0.216988</td>\n",
       "      <td>0.070509</td>\n",
       "      <td>1.932144e-02</td>\n",
       "      <td>6</td>\n",
       "      <td>168217032</td>\n",
       "      <td>168220262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>310</td>\n",
       "      <td>...</td>\n",
       "      <td>168100952</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.102941</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ENSG00000287056</td>\n",
       "      <td>4.858106e-02</td>\n",
       "      <td>4.858106e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000144130-chr2_112830976_A_G</th>\n",
       "      <td>chr2_112830976_A_G</td>\n",
       "      <td>0.017801</td>\n",
       "      <td>-0.395662</td>\n",
       "      <td>0.167722</td>\n",
       "      <td>1.944629e-02</td>\n",
       "      <td>2</td>\n",
       "      <td>112721020</td>\n",
       "      <td>112742879</td>\n",
       "      <td>NT5DC4</td>\n",
       "      <td>310</td>\n",
       "      <td>...</td>\n",
       "      <td>112830976</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.073529</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ENSG00000144130</td>\n",
       "      <td>4.883123e-02</td>\n",
       "      <td>4.883123e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000119913-chr10_112484985_A_C</th>\n",
       "      <td>chr10_112484985_A_C</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>-0.378626</td>\n",
       "      <td>0.103072</td>\n",
       "      <td>1.984266e-02</td>\n",
       "      <td>10</td>\n",
       "      <td>112283400</td>\n",
       "      <td>112305038</td>\n",
       "      <td>TECTB</td>\n",
       "      <td>310</td>\n",
       "      <td>...</td>\n",
       "      <td>112484985</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.674815</td>\n",
       "      <td>ENSG00000119913</td>\n",
       "      <td>4.976167e-02</td>\n",
       "      <td>4.976167e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2470 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  snp_id   p_value      beta  \\\n",
       "QTL                                                                            \n",
       "ENSG00000269376-chr13_112983450_T_C  chr13_112983450_T_C  0.000066 -0.364427   \n",
       "ENSG00000183463-chr13_27824764_C_T    chr13_27824764_C_T  0.047441 -0.089739   \n",
       "ENSG00000259182-chr15_101221494_A_C  chr15_101221494_A_C  0.202509 -0.088910   \n",
       "ENSG00000286922-chr12_127567286_A_G  chr12_127567286_A_G  0.265404 -0.030918   \n",
       "ENSG00000286246-chr12_129375924_C_G  chr12_129375924_C_G  0.949835 -0.000101   \n",
       "...                                                  ...       ...       ...   \n",
       "ENSG00000250842-chr5_145190494_A_T    chr5_145190494_A_T  0.000033  0.535958   \n",
       "ENSG00000267537-chr19_28873752_G_A    chr19_28873752_G_A  0.000285 -0.620958   \n",
       "ENSG00000287056-chr6_168100952_C_G    chr6_168100952_C_G  0.001926 -0.216988   \n",
       "ENSG00000144130-chr2_112830976_A_G    chr2_112830976_A_G  0.017801 -0.395662   \n",
       "ENSG00000119913-chr10_112484985_A_C  chr10_112484985_A_C  0.000204 -0.378626   \n",
       "\n",
       "                                      beta_se  empirical_feature_p_value  \\\n",
       "QTL                                                                        \n",
       "ENSG00000269376-chr13_112983450_T_C  0.092665               0.000000e+00   \n",
       "ENSG00000183463-chr13_27824764_C_T   0.045430               0.000000e+00   \n",
       "ENSG00000259182-chr15_101221494_A_C  0.069866               0.000000e+00   \n",
       "ENSG00000286922-chr12_127567286_A_G  0.027792               0.000000e+00   \n",
       "ENSG00000286246-chr12_129375924_C_G  0.001608              4.101945e-163   \n",
       "...                                       ...                        ...   \n",
       "ENSG00000250842-chr5_145190494_A_T   0.130797               1.828702e-02   \n",
       "ENSG00000267537-chr19_28873752_G_A   0.172965               1.854836e-02   \n",
       "ENSG00000287056-chr6_168100952_C_G   0.070509               1.932144e-02   \n",
       "ENSG00000144130-chr2_112830976_A_G   0.167722               1.944629e-02   \n",
       "ENSG00000119913-chr10_112484985_A_C  0.103072               1.984266e-02   \n",
       "\n",
       "                                     feature_chromosome  feature_start  \\\n",
       "QTL                                                                      \n",
       "ENSG00000269376-chr13_112983450_T_C                  13      113009671   \n",
       "ENSG00000183463-chr13_27824764_C_T                   13       27977717   \n",
       "ENSG00000259182-chr15_101221494_A_C                  15      101168530   \n",
       "ENSG00000286922-chr12_127567286_A_G                  12      127486938   \n",
       "ENSG00000286246-chr12_129375924_C_G                  12      129622929   \n",
       "...                                                 ...            ...   \n",
       "ENSG00000250842-chr5_145190494_A_T                    5      145337932   \n",
       "ENSG00000267537-chr19_28873752_G_A                   19       28602379   \n",
       "ENSG00000287056-chr6_168100952_C_G                    6      168217032   \n",
       "ENSG00000144130-chr2_112830976_A_G                    2      112721020   \n",
       "ENSG00000119913-chr10_112484985_A_C                  10      112283400   \n",
       "\n",
       "                                     feature_end gene_name  n_samples  ...  \\\n",
       "QTL                                                                    ...   \n",
       "ENSG00000269376-chr13_112983450_T_C    113010319       NaN        277  ...   \n",
       "ENSG00000183463-chr13_27824764_C_T      27988693      URAD        277  ...   \n",
       "ENSG00000259182-chr15_101221494_A_C    101170821       NaN        277  ...   \n",
       "ENSG00000286922-chr12_127567286_A_G    127533242       NaN        277  ...   \n",
       "ENSG00000286246-chr12_129375924_C_G    129625366       NaN        277  ...   \n",
       "...                                          ...       ...        ...  ...   \n",
       "ENSG00000250842-chr5_145190494_A_T     145381670       NaN        310  ...   \n",
       "ENSG00000267537-chr19_28873752_G_A      28648303       NaN        310  ...   \n",
       "ENSG00000287056-chr6_168100952_C_G     168220262       NaN        310  ...   \n",
       "ENSG00000144130-chr2_112830976_A_G     112742879    NT5DC4        310  ...   \n",
       "ENSG00000119913-chr10_112484985_A_C    112305038     TECTB        310  ...   \n",
       "\n",
       "                                     snp_position  assessed_allele  call_rate  \\\n",
       "QTL                                                                             \n",
       "ENSG00000269376-chr13_112983450_T_C     112983450                C        1.0   \n",
       "ENSG00000183463-chr13_27824764_C_T       27824764                C        1.0   \n",
       "ENSG00000259182-chr15_101221494_A_C     101221494                A        1.0   \n",
       "ENSG00000286922-chr12_127567286_A_G     127567286                G        1.0   \n",
       "ENSG00000286246-chr12_129375924_C_G     129375924                C        1.0   \n",
       "...                                           ...              ...        ...   \n",
       "ENSG00000250842-chr5_145190494_A_T      145190494                A        1.0   \n",
       "ENSG00000267537-chr19_28873752_G_A       28873752                G        1.0   \n",
       "ENSG00000287056-chr6_168100952_C_G      168100952                C        1.0   \n",
       "ENSG00000144130-chr2_112830976_A_G      112830976                G        1.0   \n",
       "ENSG00000119913-chr10_112484985_A_C     112484985                A        1.0   \n",
       "\n",
       "                                          maf     hwe_p       feature_id  \\\n",
       "QTL                                                                        \n",
       "ENSG00000269376-chr13_112983450_T_C  0.044776  1.000000  ENSG00000269376   \n",
       "ENSG00000183463-chr13_27824764_C_T   0.089552  0.073438  ENSG00000183463   \n",
       "ENSG00000259182-chr15_101221494_A_C  0.044776  0.110206  ENSG00000259182   \n",
       "ENSG00000286922-chr12_127567286_A_G  0.074627  1.000000  ENSG00000286922   \n",
       "ENSG00000286246-chr12_129375924_C_G  0.089552  0.073438  ENSG00000286246   \n",
       "...                                       ...       ...              ...   \n",
       "ENSG00000250842-chr5_145190494_A_T   0.316176  1.000000  ENSG00000250842   \n",
       "ENSG00000267537-chr19_28873752_G_A   0.066176  0.246147  ENSG00000267537   \n",
       "ENSG00000287056-chr6_168100952_C_G   0.102941  1.000000  ENSG00000287056   \n",
       "ENSG00000144130-chr2_112830976_A_G   0.073529  1.000000  ENSG00000144130   \n",
       "ENSG00000119913-chr10_112484985_A_C  0.176471  0.674815  ENSG00000119913   \n",
       "\n",
       "                                    global_corrected_pValue  \\\n",
       "QTL                                                           \n",
       "ENSG00000269376-chr13_112983450_T_C            0.000000e+00   \n",
       "ENSG00000183463-chr13_27824764_C_T             0.000000e+00   \n",
       "ENSG00000259182-chr15_101221494_A_C            0.000000e+00   \n",
       "ENSG00000286922-chr12_127567286_A_G            0.000000e+00   \n",
       "ENSG00000286246-chr12_129375924_C_G           1.930191e-161   \n",
       "...                                                     ...   \n",
       "ENSG00000250842-chr5_145190494_A_T             4.610052e-02   \n",
       "ENSG00000267537-chr19_28873752_G_A             4.669823e-02   \n",
       "ENSG00000287056-chr6_168100952_C_G             4.858106e-02   \n",
       "ENSG00000144130-chr2_112830976_A_G             4.883123e-02   \n",
       "ENSG00000119913-chr10_112484985_A_C            4.976167e-02   \n",
       "\n",
       "                                     global_corrected_pValue_BH  \\\n",
       "QTL                                                               \n",
       "ENSG00000269376-chr13_112983450_T_C                0.000000e+00   \n",
       "ENSG00000183463-chr13_27824764_C_T                 0.000000e+00   \n",
       "ENSG00000259182-chr15_101221494_A_C                0.000000e+00   \n",
       "ENSG00000286922-chr12_127567286_A_G                0.000000e+00   \n",
       "ENSG00000286246-chr12_129375924_C_G               2.223254e-161   \n",
       "...                                                         ...   \n",
       "ENSG00000250842-chr5_145190494_A_T                 4.610052e-02   \n",
       "ENSG00000267537-chr19_28873752_G_A                 4.669823e-02   \n",
       "ENSG00000287056-chr6_168100952_C_G                 4.858106e-02   \n",
       "ENSG00000144130-chr2_112830976_A_G                 4.883123e-02   \n",
       "ENSG00000119913-chr10_112484985_A_C                4.976167e-02   \n",
       "\n",
       "                                     global_corrected_pValue_BF  \\\n",
       "QTL                                                               \n",
       "ENSG00000269376-chr13_112983450_T_C                0.000000e+00   \n",
       "ENSG00000183463-chr13_27824764_C_T                 0.000000e+00   \n",
       "ENSG00000259182-chr15_101221494_A_C                0.000000e+00   \n",
       "ENSG00000286922-chr12_127567286_A_G                0.000000e+00   \n",
       "ENSG00000286246-chr12_129375924_C_G               1.111627e-160   \n",
       "...                                                         ...   \n",
       "ENSG00000250842-chr5_145190494_A_T                 1.000000e+00   \n",
       "ENSG00000267537-chr19_28873752_G_A                 1.000000e+00   \n",
       "ENSG00000287056-chr6_168100952_C_G                 1.000000e+00   \n",
       "ENSG00000144130-chr2_112830976_A_G                 1.000000e+00   \n",
       "ENSG00000119913-chr10_112484985_A_C                1.000000e+00   \n",
       "\n",
       "                                               celltype  \n",
       "QTL                                                      \n",
       "ENSG00000269376-chr13_112983450_T_C         immature_EN  \n",
       "ENSG00000183463-chr13_27824764_C_T          immature_EN  \n",
       "ENSG00000259182-chr15_101221494_A_C         immature_EN  \n",
       "ENSG00000286922-chr12_127567286_A_G         immature_EN  \n",
       "ENSG00000286246-chr12_129375924_C_G         immature_EN  \n",
       "...                                                 ...  \n",
       "ENSG00000250842-chr5_145190494_A_T   Differentiating_RG  \n",
       "ENSG00000267537-chr19_28873752_G_A   Differentiating_RG  \n",
       "ENSG00000287056-chr6_168100952_C_G   Differentiating_RG  \n",
       "ENSG00000144130-chr2_112830976_A_G   Differentiating_RG  \n",
       "ENSG00000119913-chr10_112484985_A_C  Differentiating_RG  \n",
       "\n",
       "[2470 rows x 25 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtls = pd.read_csv(PRECOMPUTED_EQTLS_TSV,\n",
    "                   sep='\\t',\n",
    "                   header=0,\n",
    "                   index_col=21)\n",
    "eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:18:00.306851Z",
     "iopub.status.busy": "2024-12-04T09:18:00.306497Z",
     "iopub.status.idle": "2024-12-04T09:18:00.313928Z",
     "shell.execute_reply": "2024-12-04T09:18:00.312939Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['immature_EN', 'Stressed_Prog', 'IPC', 'Glia', 'UL-EN',\n",
       "       'Midbrain_EN', 'Dividing_Glia', 'DL-EN', 'Interneurons',\n",
       "       'Stressed_Neurons', 'Differentiating_RG'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtls['celltype'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T09:18:00.317349Z",
     "iopub.status.busy": "2024-12-04T09:18:00.316864Z",
     "iopub.status.idle": "2024-12-04T09:18:00.458266Z",
     "shell.execute_reply": "2024-12-04T09:18:00.456951Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_chromosome</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>celltype</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>cell_type_custom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>QTL</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000269376-chr13_112983450_T_C</th>\n",
       "      <td>13</td>\n",
       "      <td>112983450</td>\n",
       "      <td>chr13_112983450_T_C</td>\n",
       "      <td>immature_EN</td>\n",
       "      <td>chr13</td>\n",
       "      <td>112983449</td>\n",
       "      <td>Neural-progenitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183463-chr13_27824764_C_T</th>\n",
       "      <td>13</td>\n",
       "      <td>27824764</td>\n",
       "      <td>chr13_27824764_C_T</td>\n",
       "      <td>immature_EN</td>\n",
       "      <td>chr13</td>\n",
       "      <td>27824763</td>\n",
       "      <td>Neural-progenitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000259182-chr15_101221494_A_C</th>\n",
       "      <td>15</td>\n",
       "      <td>101221494</td>\n",
       "      <td>chr15_101221494_A_C</td>\n",
       "      <td>immature_EN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101221493</td>\n",
       "      <td>Neural-progenitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286922-chr12_127567286_A_G</th>\n",
       "      <td>12</td>\n",
       "      <td>127567286</td>\n",
       "      <td>chr12_127567286_A_G</td>\n",
       "      <td>immature_EN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>127567285</td>\n",
       "      <td>Neural-progenitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000286246-chr12_129375924_C_G</th>\n",
       "      <td>12</td>\n",
       "      <td>129375924</td>\n",
       "      <td>chr12_129375924_C_G</td>\n",
       "      <td>immature_EN</td>\n",
       "      <td>chr12</td>\n",
       "      <td>129375923</td>\n",
       "      <td>Neural-progenitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000250842-chr5_145190494_A_T</th>\n",
       "      <td>5</td>\n",
       "      <td>145190494</td>\n",
       "      <td>chr5_145190494_A_T</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "      <td>chr5</td>\n",
       "      <td>145190493</td>\n",
       "      <td>Glia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000267537-chr19_28873752_G_A</th>\n",
       "      <td>19</td>\n",
       "      <td>28873752</td>\n",
       "      <td>chr19_28873752_G_A</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>28873751</td>\n",
       "      <td>Glia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000287056-chr6_168100952_C_G</th>\n",
       "      <td>6</td>\n",
       "      <td>168100952</td>\n",
       "      <td>chr6_168100952_C_G</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "      <td>chr6</td>\n",
       "      <td>168100951</td>\n",
       "      <td>Glia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000144130-chr2_112830976_A_G</th>\n",
       "      <td>2</td>\n",
       "      <td>112830976</td>\n",
       "      <td>chr2_112830976_A_G</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "      <td>chr2</td>\n",
       "      <td>112830975</td>\n",
       "      <td>Glia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000119913-chr10_112484985_A_C</th>\n",
       "      <td>10</td>\n",
       "      <td>112484985</td>\n",
       "      <td>chr10_112484985_A_C</td>\n",
       "      <td>Differentiating_RG</td>\n",
       "      <td>chr10</td>\n",
       "      <td>112484984</td>\n",
       "      <td>Glia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1953 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     snp_chromosome        end  \\\n",
       "QTL                                                              \n",
       "ENSG00000269376-chr13_112983450_T_C              13  112983450   \n",
       "ENSG00000183463-chr13_27824764_C_T               13   27824764   \n",
       "ENSG00000259182-chr15_101221494_A_C              15  101221494   \n",
       "ENSG00000286922-chr12_127567286_A_G              12  127567286   \n",
       "ENSG00000286246-chr12_129375924_C_G              12  129375924   \n",
       "...                                             ...        ...   \n",
       "ENSG00000250842-chr5_145190494_A_T                5  145190494   \n",
       "ENSG00000267537-chr19_28873752_G_A               19   28873752   \n",
       "ENSG00000287056-chr6_168100952_C_G                6  168100952   \n",
       "ENSG00000144130-chr2_112830976_A_G                2  112830976   \n",
       "ENSG00000119913-chr10_112484985_A_C              10  112484985   \n",
       "\n",
       "                                                      id            celltype  \\\n",
       "QTL                                                                            \n",
       "ENSG00000269376-chr13_112983450_T_C  chr13_112983450_T_C         immature_EN   \n",
       "ENSG00000183463-chr13_27824764_C_T    chr13_27824764_C_T         immature_EN   \n",
       "ENSG00000259182-chr15_101221494_A_C  chr15_101221494_A_C         immature_EN   \n",
       "ENSG00000286922-chr12_127567286_A_G  chr12_127567286_A_G         immature_EN   \n",
       "ENSG00000286246-chr12_129375924_C_G  chr12_129375924_C_G         immature_EN   \n",
       "...                                                  ...                 ...   \n",
       "ENSG00000250842-chr5_145190494_A_T    chr5_145190494_A_T  Differentiating_RG   \n",
       "ENSG00000267537-chr19_28873752_G_A    chr19_28873752_G_A  Differentiating_RG   \n",
       "ENSG00000287056-chr6_168100952_C_G    chr6_168100952_C_G  Differentiating_RG   \n",
       "ENSG00000144130-chr2_112830976_A_G    chr2_112830976_A_G  Differentiating_RG   \n",
       "ENSG00000119913-chr10_112484985_A_C  chr10_112484985_A_C  Differentiating_RG   \n",
       "\n",
       "                                       chr      start    cell_type_custom  \n",
       "QTL                                                                        \n",
       "ENSG00000269376-chr13_112983450_T_C  chr13  112983449  Neural-progenitors  \n",
       "ENSG00000183463-chr13_27824764_C_T   chr13   27824763  Neural-progenitors  \n",
       "ENSG00000259182-chr15_101221494_A_C  chr15  101221493  Neural-progenitors  \n",
       "ENSG00000286922-chr12_127567286_A_G  chr12  127567285  Neural-progenitors  \n",
       "ENSG00000286246-chr12_129375924_C_G  chr12  129375923  Neural-progenitors  \n",
       "...                                    ...        ...                 ...  \n",
       "ENSG00000250842-chr5_145190494_A_T    chr5  145190493                Glia  \n",
       "ENSG00000267537-chr19_28873752_G_A   chr19   28873751                Glia  \n",
       "ENSG00000287056-chr6_168100952_C_G    chr6  168100951                Glia  \n",
       "ENSG00000144130-chr2_112830976_A_G    chr2  112830975                Glia  \n",
       "ENSG00000119913-chr10_112484985_A_C  chr10  112484984                Glia  \n",
       "\n",
       "[1953 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Format df\n",
    "eqtls_bed = eqtls[['snp_chromosome', 'snp_position', 'snp_id', 'celltype']].copy()\n",
    "eqtls_bed['chr'] = 'chr' + eqtls_bed['snp_chromosome'].astype(str)\n",
    "eqtls_bed['start'] = eqtls_bed['snp_position'] - 1 # Make index 0-based open\n",
    "eqtls_bed['cell_type_custom'] = eqtls['celltype'].map(ct_map_i_alt)\n",
    "eqtls_bed = eqtls_bed.rename(columns={'snp_position': 'end', 'snp_id': 'id'})\n",
    "\n",
    "# Filter out 'Discard' cell-types\n",
    "\n",
    "eqtls_bed = eqtls_bed[~(eqtls_bed['cell_type_custom'] == 'Discard')].copy()\n",
    "\n",
    "# Save bed for all cell-types\n",
    "eqtls_bed_all_cts = eqtls_bed[['chr', 'start', 'end', 'id']].sort_values(by=['chr', 'start'], ascending=[True, True])\n",
    "\n",
    "eqtls_bed_all_cts_path = f'{EQTLS_DIR}/{CT_MAP_ID}/eqtls_all_cell-types.bed'\n",
    "create_dir(eqtls_bed_all_cts_path)\n",
    "eqtls_bed_all_cts.to_csv(eqtls_bed_all_cts_path, sep='\\t', header=False, index=False)\n",
    "eqtls_bed_all_cts\n",
    "\n",
    "# Save bed for individual cell-types\n",
    "for ct in eqtls_bed['cell_type_custom'].unique():\n",
    "\n",
    "    eqtls_bed_ct = eqtls_bed[eqtls_bed['cell_type_custom'] == ct].copy()\n",
    "    eqtls_bed_ct = eqtls_bed_ct[['chr', 'start', 'end', 'id']].sort_values(by=['chr', 'start'], ascending=[True, True])\n",
    "\n",
    "    eqtls_bed_ct_path = f'{EQTLS_DIR}/{CT_MAP_ID}/{ct}/eqtls.bed'\n",
    "    create_dir(eqtls_bed_ct_path)\n",
    "    eqtls_bed_ct.to_csv(eqtls_bed_ct_path, sep='\\t', header=False, index=False)\n",
    "    \n",
    "eqtls_bed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "main05",
   "language": "python",
   "name": "main05"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
